### ISSUE TITLE
selectLeft reduces selection instead of increasing

### ISSUE THREAD

**Issue by bronson** [atom/atom#6928]
In this Chrome Devtools session, I call selectLeft() and the selection _decreases_ in size.

The [selectLeft documentation](https://atom.io/docs/api/v0.201.0/Selection#instance-selectLeft) implies that this should not happen:

    Selects the text one position left of the cursor.

Is this a bug in selectLeft?  Is my selection inconsistent?  (it was created by some complex vim-mode code).  Something else?

![screen shot 2015-05-24 at 12 05 48 am](https://cloud.githubusercontent.com/assets/1776/7787027/b0832336-01a8-11e5-8aa8-1f488288f910.png)

**Comment by bronson** [comment]
aha, [reversed selections are legal](https://atom.io/docs/api/v0.201.0/Selection#instance-isReversed)

So selectLeft() doesn't take reversed selections into account.  That's that's a bug, yeah?

**Comment by bronson** [comment]
My final guess: not a bug, just incomplete documentation.

From a UI standpoint, `moveLeft()` and `moveRight()` make sense if you picture them being in response to hitting <kbd>shift-left</kbd> or <kbd>shift-right</kbd> while a selection is active.  However, that means that, if the selection is reversed, then selectLeft() will DESELECT the text one position to the left of the cursor.

If this is right, happy to submit a PR to add "If the selection is reversed, deselects one position left of the cursor" etc. to the docs.  That knowledge would have saved me a bit of time tonight...

**Comment by izuzak** [comment]
> That knowledge would have saved me a bit of time tonight.

Sorry to hear that, @bronson, and thanks for taking the time to provide this feedback. :bow:

I think you're right and your last comment explains why this happens, but I'm wondering if there still might be something to fix here in Atom -- @nathansobo @maxbrunsfeld @benogle, could you take a quick look perhaps? 

I expected different behavior than the one that's currently in Atom. Here's a GIF which demonstrates this:

![selection](https://cloud.githubusercontent.com/assets/38924/7787359/1873d10c-0208-11e5-86b0-8e4654e077ac.gif)

I double-click a word to select it, and then `shift-left` to select one character to the left. My expectation was that the selection would be extended to the left instead of reduced on the right. (@bronson's comment above explains why this happens.)

For example, Chrome and OSX Finder behaved as I expected in that situation:

![selection2](https://cloud.githubusercontent.com/assets/38924/7787369/82bf9a1e-0208-11e5-8aee-67b851e7a256.gif)

In other words, when a selection is created -- I expect Atom to "wait" and see in which direction I start expanding, and then use that for further expanding the selection. I don't expect the selection to shrink on the initial expansion.

I think that's how moving a cursor already works with a selection -- here's what happens when I double-click a word and then just press `left` to move the cursor (notice that it's positioned to the left of the selection):

![selection3](https://cloud.githubusercontent.com/assets/38924/7787388/cbb5a050-0209-11e5-874c-908d58ad30c7.gif)

**Comment by maxbrunsfeld** [comment]
Yeah this is the expected behavior. @bronson Your proposed change to the docs sounds good; I'd definitely merge such a PR.

**Comment by nathansobo** [comment]
The main issue isn't a bug, but what @izuzak points out could sort of be considered a bug. On some other code paths we use `Selection::initialScreenRange` to wait to "decide" on the directionality of the selection if it's created via a double-click or triple-click. This is why double- or triple-clicking works correctly when followed by a drag. We would need to use the `::initialScreenRange` in other methods on the selection in order to generalize this behavior.

**Comment by izuzak** [comment]
> The main issue isn't a bug, but what @izuzak points out could sort of be considered a bug. On some other code paths we use Selection::initialScreenRange to wait to "decide" on the directionality of the selection if it's created via a double-click or triple-click.

Just noticed that this is reported in another issue (https://github.com/atom/atom/issues/6485), so I'm going to close this in favor of that.
